testcases/cve/stack_clash.c

```diff
@@ -0,0 +1,242 @@
+ * considered hard to hop above. Code adapted from the Novell's bugzilla [2].
+ *
+ * The code `mmap(2)`s region close to the stack end. The code then allocates
+ * memory on stack until it hits guard page and SIGSEGV or SIGBUS is generated
+ * by the kernel. The signal handler checks that fault address is further than
+ * THRESHOLD from the mmapped area.
+ *
+ * We read /proc/self/maps to examine exact top of the stack and `mmap(2)`
+ * our region exactly GAP_PAGES * PAGE_SIZE away.
+ *
+ * [1] https://blog.qualys.com/securitylabs/2017/06/19/the-stack-clash
+ * [2] https://bugzilla.novell.com/show_bug.cgi?id=CVE-2017-1000364
+ */
+
+#include <sys/mman.h>
+#include <sys/wait.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <alloca.h>
+#include <signal.h>
+#include <stdlib.h>
+
+#include "tst_test.h"
+#include "tst_safe_stdio.h"
+
+static unsigned long PAGE_SIZE;
+static unsigned long PAGE_MASK;
+static unsigned long GAP_PAGES = 256;
+static unsigned long THRESHOLD;
+static int STACK_GROWSDOWN;
+
+#define SIGNAL_STACK_SIZE	(1UL<<20)
+#define FRAME_SIZE		1024
+
+#define EXIT_TESTBROKE		TBROK
+
+void exhaust_stack_into_sigsegv(void)
+{
+	volatile char * ptr = alloca(FRAME_SIZE - sizeof(long));
+	*ptr = '\0';
+	exhaust_stack_into_sigsegv();
+}
+
+#define MAPPED_LEN PAGE_SIZE
+static unsigned long mapped_addr;
+
+void segv_handler(int sig, siginfo_t *info, void *data LTP_ATTRIBUTE_UNUSED)
+{
+	unsigned long fault_addr = (unsigned long)info->si_addr;
+	unsigned long mmap_end = mapped_addr + MAPPED_LEN;
+	ssize_t diff;
+
+	if (sig != SIGSEGV)
+		return;
+
+	if (STACK_GROWSDOWN)
+		diff = fault_addr - mmap_end;
+	else
+		diff = mapped_addr - fault_addr;
+
+	tst_res(TINFO,
+		"mmap = [%lx, %lx), addr = %lx, diff = %lx, THRESHOLD = %lx",
+		mapped_addr, mmap_end, fault_addr, diff, THRESHOLD);
+	if (diff < 0 || (unsigned long)diff < THRESHOLD)
+		_exit(EXIT_FAILURE);
+	else
+		_exit(EXIT_SUCCESS);
+}
+
+unsigned long read_stack_addr_from_proc(void)
+{
+	FILE *fh;
+	char buf[1024];
+	unsigned long stack_top = -1UL, start, end;
+
+	fh = SAFE_FOPEN("/proc/self/maps", "r");
+
+	while (!feof(fh)) {
+		if (fgets(buf, sizeof(buf), fh) == NULL) {
+			tst_brk(TBROK | TERRNO, "fgets");
+			goto out;
+		}
+
+		if (!strstr(buf, "[stack"))
+			continue;
+
+		if (sscanf(buf, "%lx-%lx", &start, &end) != 2) {
+			tst_brk(TBROK | TERRNO, "sscanf");
+			goto out;
+		}
+
+		if (STACK_GROWSDOWN)
+			stack_top = start;
+		else
+			stack_top = end;
+		break;
+	}
+
+out:
+	SAFE_FCLOSE(fh);
+	return stack_top;
+}
+
+void do_child(void)
+{
+	unsigned long stack_addr;
+	stack_t signal_stack;
+	struct sigaction segv_sig = {.sa_sigaction = segv_handler, .sa_flags = SA_ONSTACK|SA_SIGINFO};
+	void *map;
+	unsigned long gap = GAP_PAGES * PAGE_SIZE;
+
+	stack_addr = read_stack_addr_from_proc();
+	if (stack_addr == -1UL) {
+		tst_brk(TBROK, "can't read stack top from /proc/self/maps");
+		return;
+	}
+
+	if (STACK_GROWSDOWN)
+		mapped_addr = stack_addr - gap - MAPPED_LEN;
```

Comment:
So, the stack shouldn't be allowed grow at all, and we can presumably ignore stack limit for unprivileged user.

The only (extremely unlikely, let's ignore this) way I see this going wrong is if your stack limit on old kernel is so low, you can't grow your stack at all:
```
$ 4.11.0 (ppc64le)
$ ulimit -s 64
$ ./stack_clash 
tst_test.c:898: INFO: Timeout per run is 0h 05m 00s
stack_clash.c:200: INFO: STACK_GROWSDOWN = 1 == 0x3ffff8eac1a0 > 0x3ffff8eac110
stack_clash.c:147: INFO: Stack top:0x3ffff8ea0000 mmap:0x3ffff7e90000+0x10000
stack_clash.c:82: INFO: mmap = [3ffff7e90000, 3ffff7ea0000), addr = 3ffff8e9fd30, diff = fffd30, THRESHOLD = ff0000
stack_clash.c:225: PASS: stack is far enough from mmaped area
```